crates/uv-pypi-types/src/conflicting_groups.rs

```diff
@@ -0,0 +1,192 @@
+
+    fn try_from(groups: Vec<ConflictingGroup>) -> Result<ConflictingGroups, ConflictingGroupError> {
+        if groups.len() < 2 {
+            return Err(ConflictingGroupError::NotEnoughGroups { len: groups.len() });
+        }
+        Ok(ConflictingGroups(groups))
+    }
+}
+
+/// A single item in a set conflicting groups.
+///
+/// Each item is a pair of a package and a corresponding extra name for that
+/// package.
+#[derive(
+    Debug,
+    Default,
+    Clone,
+    Eq,
+    Hash,
+    PartialEq,
+    PartialOrd,
+    Ord,
+    serde::Deserialize,
+    serde::Serialize,
+    schemars::JsonSchema,
+)]
+pub struct ConflictingGroup {
+    package: PackageName,
+    extra: ExtraName,
+}
+
+impl ConflictingGroup {
+    /// Returns the package name of this conflicting group.
+    pub fn package(&self) -> &PackageName {
+        &self.package
+    }
+
+    /// Returns the extra name of this conflicting group.
+    pub fn extra(&self) -> &ExtraName {
+        &self.extra
+    }
+
+    /// Returns this group as a new type with its fields borrowed.
+    pub fn as_ref(&self) -> ConflictingGroupRef<'_> {
+        ConflictingGroupRef {
+            package: self.package(),
+            extra: self.extra(),
+        }
+    }
+}
+
+impl From<(PackageName, ExtraName)> for ConflictingGroup {
+    fn from((package, extra): (PackageName, ExtraName)) -> ConflictingGroup {
+        ConflictingGroup { package, extra }
+    }
+}
+
+/// A single item in a set conflicting groups, by reference.
+///
+/// Each item is a pair of a package and a corresponding extra name for that
+/// package.
+#[derive(Debug, Clone, Copy, Eq, Hash, PartialEq, PartialOrd, Ord)]
+pub struct ConflictingGroupRef<'a> {
+    package: &'a PackageName,
+    extra: &'a ExtraName,
+}
+
+impl<'a> ConflictingGroupRef<'a> {
+    /// Returns the package name of this conflicting group.
+    pub fn package(&self) -> &'a PackageName {
+        self.package
+    }
+
+    /// Returns the extra name of this conflicting group.
+    pub fn extra(&self) -> &'a ExtraName {
+        self.extra
+    }
+
+    /// Converts this borrowed conflicting group to its owned variant.
+    pub fn to_owned(&self) -> ConflictingGroup {
+        ConflictingGroup {
+            package: self.package().clone(),
+            extra: self.extra().clone(),
+        }
+    }
+}
+
+impl<'a> From<(&'a PackageName, &'a ExtraName)> for ConflictingGroupRef<'a> {
+    fn from((package, extra): (&'a PackageName, &'a ExtraName)) -> ConflictingGroupRef<'a> {
+        ConflictingGroupRef { package, extra }
+    }
+}
+
+/// An error that occurs when the given conflicting groups are invalid somehow.
+#[derive(Debug, thiserror::Error)]
+pub enum ConflictingGroupError {
+    /// An error for when there are too few conflicting groups.
+    ///
+    /// (A singleton or empty set of groups doesn't make sense.)
+    #[error("Each set of conflicting groups must have at least 2 entries, but found only {len}")]
```

Comment:
Nit: "two" instead of "2", sorry!

Reply:
Also might be worth special-casing zero instead of saying "but found only 0".

Reply:
Yeah, I totally get that style. I ended up using `2` intentionally for reasons of "parallelism," since "but found only {len}" would also be written numerically. So I think it would be weird to, e.g., have `at least two entries, but found only 1`.

However, if we're special casing the empty case to make it read a little nicer, we might as well special case the singleton case since those are the only two error cases related to the number of groups. In which case, we can just write the number out in English. :P 

So... long winded way of saying, SGTM. Lol.